src/herder/TxSetUtils.cpp

```diff
@@ -0,0 +1,393 @@
+        }
+        if (tx2 == nullptr || tx2->empty())
+        {
+            return false;
+        }
+
+        auto& top1 = tx1->front();
+        auto& top2 = tx2->front();
+
+        auto cmp3 = feeRate3WayCompare(top1, top2);
+
+        if (cmp3 != 0)
+        {
+            return cmp3 < 0;
+        }
+        // use hash of transaction as a tie breaker
+        return lessThanXored(top1->getFullHash(), top2->getFullHash(), mSeed);
+    }
+};
+
+TxSetFrameConstPtr
+surgePricingFilter(TxSetFrameConstPtr txSet, Application& app)
+{
+    ZoneScoped;
+    LedgerTxn ltx(app.getLedgerTxnRoot());
+    auto header = ltx.loadHeader();
+
+    bool maxIsOps = protocolVersionStartsFrom(header.current().ledgerVersion,
+                                              ProtocolVersion::V_11);
+
+    size_t opsLeft = app.getLedgerManager().getLastMaxTxSetSizeOps();
+
+    auto curSizeOps =
+        maxIsOps ? txSet->sizeOp() : (txSet->sizeTx() * MAX_OPS_PER_TX);
+    if (curSizeOps > opsLeft)
+    {
+        CLOG_WARNING(Herder, "surge pricing in effect! {} > {}", curSizeOps,
+                     opsLeft);
+
+        auto actTxQueueMap = buildAccountTxQueues(*txSet);
+
+        std::priority_queue<TxSetFrame::AccountTransactionQueue*,
+                            std::vector<TxSetFrame::AccountTransactionQueue*>,
+                            SurgeCompare>
+            surgeQueue;
+
+        for (auto& am : actTxQueueMap)
+        {
+            surgeQueue.push(&am.second);
+        }
+
+        TxSetFrame::Transactions updatedSet;
+        updatedSet.reserve(txSet->sizeTx());
+        while (opsLeft > 0 && !surgeQueue.empty())
+        {
+            auto cur = surgeQueue.top();
+            surgeQueue.pop();
+            // inspect the top candidate queue
+            auto& curTopTx = cur->front();
+            size_t opsCount =
+                maxIsOps ? curTopTx->getNumOperations() : MAX_OPS_PER_TX;
+            if (opsCount <= opsLeft)
+            {
+                // pop from this one
+                updatedSet.emplace_back(curTopTx);
+                cur->pop_front();
+                opsLeft -= opsCount;
+                // if there are more transactions, put it back
+                if (!cur->empty())
+                {
+                    surgeQueue.push(cur);
+                }
+            }
+            else
+            {
+                // drop this transaction -> we need to drop the others
+                cur->clear();
+            }
+        }
+
+        return std::make_shared<TxSetFrame const>(txSet->previousLedgerHash(),
+                                                  updatedSet);
+    }
+    return txSet;
+}
+
+TxSetFrame::Transactions
+getInvalidTxList(Application& app, TxSetFrame const& txSet,
+                 uint64_t lowerBoundCloseTimeOffset,
+                 uint64_t upperBoundCloseTimeOffset,
+                 bool returnEarlyOnFirstInvalidTx)
+{
+    ZoneScoped;
+    LedgerTxn ltx(app.getLedgerTxnRoot());
```

Comment:
I'm surprised that we don't create a "read only" LedgerTxn here - @marta-lokhova did we miss this earlier?

Reply:
not following this comment, can you clarify? I haven't been reviewing this PR, so I need context.

Reply:
I thought we were supposed to use `TransactionMode::READ_ONLY_WITHOUT_SQL_TXN` in places like this

Reply:
We're currently using `TransactionMode::READ_ONLY_WITHOUT_SQL_TXN` on the transaction processing path (there's a significant perf impact because that path is called hundreds of times per second). We could call it on this path as well, though I'm not sure we'll see the benefits (we call this function for unique SCP messages it looks like)

Reply:
yeah perf wise would not have much impact, I was thinking about this as a strong signal that nothing changes and we're performing read only operations as in, the code ends up being easier for people to understand (and impossible to mess up)

Reply:
I think makes sense. I've updated it.